src/lib/operators/case_operator.cpp

```diff
@@ -0,0 +1,226 @@
+#include "case_operator.hpp"
+
+#include <sstream>
+
+#include "storage/column_iterables/any_column_iterator.hpp"
+#include "storage/create_iterable_from_column.hpp"
+#include "storage/materialize.hpp"
+
+namespace {
+
+using namespace opossum;  // NOLINT
+
+/**
+ * Turns a PhysicalCaseExpressionResult - which is either a THEN result or an ELSE result - into a value/null - pair
+ */
+template <typename ResultDataType>
+std::pair<bool, ResultDataType> materialize_case_result(
+    const PhysicalCaseExpressionResult<ResultDataType>& case_result, const Chunk& chunk, const ChunkOffset chunk_offset,
+    std::vector<std::optional<std::vector<std::pair<bool, ResultDataType>>>>& materialized_column_cache) {
+  /**
+   * CaseExpressionResult can be either Null, a value from a column or a constant value
+   */
+
+  if (case_result.type() == typeid(Null)) {
+    return {true, {}};
+  } else if (case_result.type() == typeid(ColumnID)) {
+    // If not yet done, materialize the column as values and nulls and cache it.
+    const auto column_id = boost::get<ColumnID>(case_result);
+    auto& materialized_column = materialized_column_cache[column_id];
+    if (!materialized_column) {
+      materialized_column.emplace();
+      materialized_column->reserve(chunk.size());
+      materialize_values_and_nulls(*chunk.get_column(column_id), *materialized_column);
```

Comment:
`materialize_column`? Also, why do you need an `in` argument? Can’t you return the vector and move it into the cache?

Reply:
Moritz and I talked about this before. The general idea was to keep the interface generic, so that both `pmr_concurrent_vector` and `pmr_vector` or whatever could be used. Also, one might want to materialize a column by appending it to an already existing container.
I’m still not very happy and I would suggest passing an iterator instead. AFAIK, this would not make it any slower and would have the advantage that anything from vectors to sets (for example getting all unique values) could be used.